### PR TITLE
CATALOG-2925: ImmutableCatalog and MutableCatalog 

### LINKED ISSUE #2925: Proposal: Change the catalog to two parts: ImmutableCatalog and MutableCatalog
**Summary**

## Background

Now, the catelog has two catagory:
* upper layer: this is readonly layer, now only one `system` database, we can not register another one, such as `information_schema` database
* bottom layer: this is metastore hosted layer, user can create/drop database/table, for now it's only one type:MetaCatalog, all the Catalog trait is following the MetaCatalog logic, we can not register another one like `github` database to it.

This makes the catalog is more flexible, a user case is:https://github.com/datafuselabs/databend/pull/2897#issuecomment-974612473

## Proposal

Change the catalog to two parts:
* ImmutableCatalog: a read-only type database catalog collection, like `system` and `information_schema` etc system who requires the read-only can be registered to the ImmutableCatalog
* MutableCatalog: a self-host type database catalog collection, like `fuse type database` and `github type database` can be registered to the MutableCatalog

### PR DESCRIPTION
**Pull request by BohuTANG**
I hereby agree to the terms of the CLA available at: https://databend.rs/policies/cla/

## Summary

Summary about this PR:
1. Strip the `Database` trait from the `Catalog` into a seprate trait, now `Catalog` trait only related to the database operations.
2. Introduce two catalogs for the `Database`: `ImmutableCatalog` and `MutableCatalog`, for each catalog we can build difference database trait object by the engine.

TODO(will do in another patch):
1. Add database register for `ImmutableCatalog` and `MutableCatalog`
2. Remove database name for some table trait methods
3. Add more unit tests

## Changelog

- Improvement

## Related Issues

Fixes #2925

## Test Plan

Unit Tests

Stateless Tests

### CODE REVIEW

**Comment by BohuTANG** [comment]
Almost done, but there is an issue blocked from the test:
```
cannot execute `LocalPool` executor from within another executor: EnterError
thread 'interpreters::interpreter_insert_into_test::test_insert_into_interpreter' panicked at 'cannot execute `LocalPool` executor from within another executor: EnterError', /home/bohu/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-executor-0.3.17/src/local_pool.rs:81:26
stack backtrace:
   0: rust_begin_unwind
             at /rustc/b69fe57261086e70aea9d5b58819a1794bf7c121/library/std/src/panicking.rs:517:5
   1: core::panicking::panic_fmt
             at /rustc/b69fe57261086e70aea9d5b58819a1794bf7c121/library/core/src/panicking.rs:96:14
   2: core::result::unwrap_failed
             at /rustc/b69fe57261086e70aea9d5b58819a1794bf7c121/library/core/src/result.rs:1617:5
   3: core::result::Result<T,E>::expect
             at /rustc/b69fe57261086e70aea9d5b58819a1794bf7c121/library/core/src/result.rs:1259:23
   4: futures_executor::local_pool::run_executor
             at /home/bohu/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-executor-0.3.17/src/local_pool.rs:81:18
   5: futures_executor::local_pool::block_on
             at /home/bohu/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-executor-0.3.17/src/local_pool.rs:315:5
   6: databend_query::sessions::context::DatabendQueryContext::build_table_from_source_plan
             at ./src/sessions/context.rs:98:13
   7: <databend_query::interpreters::plan_do_readsource::PlanDoReadSource as common_planners::plan_rewriter::PlanRewriter>::rewrite_read_data_source::{{closure}}
             at ./src/interpreters/plan_do_readsource.rs:46:25
   8: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/b69fe57261086e70aea9d5b58819a1794bf7c121/library/core/src/future/mod.rs:80:19
   9: futures_executor::local_pool::block_on::{{closure}}
             at /home/bohu/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-executor-0.3.17/src/local_pool.rs:315:23
  10: futures_executor::local_pool::run_executor::{{closure}}
             at /home/bohu/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-executor-0.3.17/src/local_pool.rs:90:37
  11: std::thread::local::LocalKey<T>::try_with
             at /rustc/b69fe57261086e70aea9d5b58819a1794bf7c121/library/std/src/thread/local.rs:399:16
  12: std::thread::local::LocalKey<T>::with
             at /rustc/b69fe57261086e70aea9d5b58819a1794bf7c121/library/std/src/thread/local.rs:375:9
  13: futures_executor::local_pool::run_executor
             at /home/bohu/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-executor-0.3.17/src/local_pool.rs:86:5
  14: futures_executor::local_pool::block_on
             at /home/bohu/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-executor-0.3.17/src/local_pool.rs:315:5
  15: <databend_query::interpreters::plan_do_readsource::PlanDoReadSource as common_planners::plan_rewriter::PlanRewriter>::rewrite_read_data_source
             at ./src/interpreters/plan_do_readsource.rs:45:9
  16: common_planners::plan_rewriter::PlanRewriter::rewrite_plan_node
             at /home/bohu/github/datafuselabs/databend/common/planners/src/plan_rewriter.rs:92:43
  17: common_planners::plan_rewriter::PlanRewriter::rewrite_projection
             at /home/bohu/github/datafuselabs/databend/common/planners/src/plan_rewriter.rs:240:25
  18: common_planners::plan_rewriter::PlanRewriter::rewrite_plan_node
             at /home/bohu/github/datafuselabs/databend/common/planners/src/plan_rewriter.rs:87:43
  19: databend_query::interpreters::utils::apply_plan_rewrite
             at ./src/interpreters/utils.rs:30:16
  20: databend_query::interpreters::interpreter_select::SelectInterpreter::schedule_query::{{closure}}
             at ./src/interpreters/interpreter_select.rs:86:30
  21: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/b69fe57261086e70aea9d5b58819a1794bf7c121/library/core/src/future/mod.rs:80:19
  22: <databend_query::interpreters::interpreter_select::SelectInterpreter as databend_query::interpreters::interpreter::Interpreter>::execute::{{closure}}::{{closure}}
             at ./src/interpreters/interpreter_select.rs:68:15
  23: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/b69fe57261086e70aea9d5b58819a1794bf7c121/library/core/src/future/mod.rs:80:19
  24: <databend_query::interpreters::interpreter_select::SelectInterpreter as databend_query::interpreters::interpreter::Interpreter>::execute::{{closure}}
             at ./src/interpreters/interpreter_select.rs:60:5
  25: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/b69fe57261086e70aea9d5b58819a1794bf7c121/library/core/src/future/mod.rs:80:19
  26: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/b69fe57261086e70aea9d5b58819a1794bf7c121/library/core/src/future/future.rs:119:9
  27: <databend_query::interpreters::interpreter_insert_into::InsertIntoInterpreter as databend_query::interpreters::interpreter::Interpreter>::execute::{{closure}}
             at ./src/interpreters/interpreter_insert_into.rs:102:17
  28: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/b69fe57261086e70aea9d5b58819a1794bf7c121/library/core/src/future/mod.rs:80:19
  29: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/b69fe57261086e70aea9d5b58819a1794bf7c121/library/core/src/future/future.rs:119:9
  30: <databend_query::interpreters::interpreter_interceptor::InterceptorInterpreter as databend_query::interpreters::interpreter::Interpreter>::execute::{{closure}}
             at ./src/interpreters/interpreter_interceptor.rs:62:29
  31: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/b69fe57261086e70aea9d5b58819a1794bf7c121/library/core/src/future/mod.rs:80:19
  32: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/b69fe57261086e70aea9d5b58819a1794bf7c121/library/core/src/future/future.rs:119:9
  33: databend_query::interpreters::interpreter_insert_into_test::test_insert_into_interpreter::{{closure}}
             at ./src/interpreters/interpreter_insert_into_test.rs:61:21
  34: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/b69fe57261086e70aea9d5b58819a1794bf7c121/library/core/src/future/mod.rs:80:19
  35: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/b69fe57261086e70aea9d5b58819a1794bf7c121/library/core/src/future/future.rs:119:9
  36: tokio::runtime::basic_scheduler::Inner<P>::block_on::{{closure}}::{{closure}}
             at /home/bohu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/runtime/basic_scheduler.rs:240:62
  37: tokio::coop::with_budget::{{closure}}
             at /home/bohu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/coop.rs:106:9
  38: std::thread::local::LocalKey<T>::try_with
             at /rustc/b69fe57261086e70aea9d5b58819a1794bf7c121/library/std/src/thread/local.rs:399:16
  39: std::thread::local::LocalKey<T>::with
             at /rustc/b69fe57261086e70aea9d5b58819a1794bf7c121/library/std/src/thread/local.rs:375:9
  40: tokio::coop::with_budget
             at /home/bohu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/coop.rs:99:5
  41: tokio::coop::budget
             at /home/bohu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/coop.rs:76:5
  42: tokio::runtime::basic_scheduler::Inner<P>::block_on::{{closure}}
             at /home/bohu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/runtime/basic_scheduler.rs:240:39
  43: tokio::runtime::basic_scheduler::enter::{{closure}}
             at /home/bohu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/runtime/basic_scheduler.rs:349:29
  44: tokio::macros::scoped_tls::ScopedKey<T>::set
             at /home/bohu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/macros/scoped_tls.rs:61:9
  45: tokio::runtime::basic_scheduler::enter
             at /home/bohu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/runtime/basic_scheduler.rs:349:5
  46: tokio::runtime::basic_scheduler::Inner<P>::block_on
             at /home/bohu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/runtime/basic_scheduler.rs:230:9
  47: tokio::runtime::basic_scheduler::InnerGuard<P>::block_on
             at /home/bohu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/runtime/basic_scheduler.rs:501:9
  48: tokio::runtime::basic_scheduler::BasicScheduler<P>::block_on
             at /home/bohu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/runtime/basic_scheduler.rs:186:24
  49: tokio::runtime::Runtime::block_on
             at /home/bohu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/runtime/mod.rs:461:46
  50: databend_query::interpreters::interpreter_insert_into_test::test_insert_into_interpreter
             at ./src/interpreters/interpreter_insert_into_test.rs:86:5
  51: databend_query::interpreters::interpreter_insert_into_test::test_insert_into_interpreter::{{closure}}
             at ./src/interpreters/interpreter_insert_into_test.rs:24:7
  52: core::ops::function::FnOnce::call_once
             at /rustc/b69fe57261086e70aea9d5b58819a1794bf7c121/library/core/src/ops/function.rs:227:5
  53: core::ops::function::FnOnce::call_once
             at /rustc/b69fe57261086e70aea9d5b58819a1794bf7c121/library/core/src/ops/function.rs:227:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: test failed, to rerun pass '--lib'

Process finished with exit code 101

```

File: common/meta/types/src/table.rs [hunk 1]
```diff
@@ -78,6 +78,9 @@ pub struct TableInfo {
     /// For a table function, it is `table_name(args)`
     pub desc: String,
 
+    /// Database name of this table.
+    pub db: String,
```
Comment (drmingdrmer) [inline common/meta/types/src/table.rs:82]: Why do we need this field?

A user may try to re-fetch a table by `db_name.table_name`, which potentially is an inconsistent operation.

File: query/src/catalogs/database.rs [hunk 2]
```diff
@@ -12,7 +12,51 @@
...
+pub trait Database: DynClone + Sync + Send {
     /// Database name.
     fn name(&self) -> &str;
+
+    // Get one table by db and table name.
+    async fn get_table(&self, db_name: &str, table_name: &str) -> Result<Arc<dyn Table>>;
+
+    async fn list_tables(&self, db_name: &str) -> Result<Vec<Arc<dyn Table>>>;
+
+    async fn get_table_meta_by_id(&self, table_id: MetaId) -> Result<(TableIdent, Arc<TableMeta>)>;
```
Comment (drmingdrmer) [inline query/src/catalogs/database.rs:40]: Is `table_id` still globally unique?

If it is, `get_table_meta_by_id` should be a method of `catalog`.

File: common/meta/types/src/table.rs (hunk 1 above)
Comment (BohuTANG) [reply common/meta/types/src/table.rs:82]: Thank you, xp!
If we move the `build_table` to catalog trait, this `db` is not used anymore, I will have a try.

File: query/src/catalogs/database.rs (hunk 2 above)
Comment (BohuTANG) [reply query/src/catalogs/database.rs:40]: The `get_table_meta_by_id` defined here, but i do not see where it used, will try move to catalog

**Comment by BohuTANG** [comment]
@drmingdrmer @dantengsky Please take a review :)

**Comment by drmingdrmer** [comment]
> The `get_table_meta_by_id` defined here, but i do not see where it used, will try move to catalog

This method is mainly meant for `commit a table`: when a CAS found a conflict `version`, `commit()` should refetch `TableMeta` and retry.
(Although the `retry` is not implemented yet)

```rust
async fn commit(
    ctx: DatabendQueryContextRef,
    table_id: MetaId,
    table_version: MetaVersion,
    new_snapshot_location: String,
) -> Result<UpsertTableOptionReply> {
    let catalog = ctx.get_catalog();
    catalog
        .upsert_table_option(UpsertTableOptionReq::new(
            &TableIdent {
                table_id,
                version: table_version,
            },
            TBL_OPT_KEY_SNAPSHOT_LOC,
            new_snapshot_location,
        ))
        .await
}

```

**Pull request merged**